### DISCUSSION TITLE
Implicit mapping does not work in ProjectTo

### DISCUSSION THREAD

**Discussion by rmja** [LuckyPennySoftware/AutoMapper#3763] [NA]
### Source/destination types

Failing test is here https://github.com/AutoMapper/AutoMapper/pull/3761

### Mapping configuration

(see failing test)

### Version: x.y.z
master

### Expected behavior
It should be possible to use implicit/explicit mapping when using ProjectTo

### Actual behavior
ProjectTo throws that no mapping is defined:
![image](https://user-images.githubusercontent.com/2112306/138241316-87e139a6-e749-40bd-8140-e69e1b12d22a.png)

### Steps to reproduce
See failing test pr. Make sure to remove the explicit mapping lines:
```
config.CreateMap<SourceValue, DestinationValue>().ConvertUsing(src => src);
config.CreateMap<SourceValue?, DestinationValue?>().ConvertUsing(src => src);
```

**Comment by lbargaoanu** [comment]
Custom conversion operators are for `Map`, not for `ProjectTo`. Try converting your test to an integration test.

**Reply by rmja** [reply]
Okay. Where exactly does the test fit then?
And, is it by design that implicit mappings to not work with ProjectTo()?

**Reply by lbargaoanu** [reply]
https://github.com/AutoMapper/AutoMapper/tree/master/src/IntegrationTests

**Reply by rmja** [reply]
I have moved the test (https://github.com/AutoMapper/AutoMapper/pull/3761/files)

**Reply by lbargaoanu** [reply]
An integration test means a test that executes a SQL statement against a real database, as the tests in that project do.

**Reply by rmja** [reply]
The thing is that it requires EF Core to register custom value converters (as far as I can tell) which is needed to actually run this through the database. The IntegrationTest project currently references EF6. Would it be ok to reference EF Core in the same project?

**Reply by lbargaoanu** [reply]
`ProjectTo` runs in the database, it cannot work with value converters. Also see https://github.com/AutoMapper/AutoMapper/discussions/3571.

**Reply by rmja** [reply]
But the thing is that it _does work_, including querying to a real database if I add:
```
config.CreateMap<SourceValue, DestinationValue>().ConvertUsing(src => src);
config.CreateMap<SourceValue?, DestinationValue?>().ConvertUsing(src => src);
```

**Reply by lbargaoanu** [reply]
If you use AM, I don't see why you need EF Core converters. If you want client evaluation, you use `Map` and if you want server evaluation, you use `ProjectTo`.

**Reply by rmja** [reply]
You are right, I do not need EF Core converters, but I need EF Core.

See the following test which fails using EF Core if the explicit mappings are not added:
https://github.com/rmja/AutoMapper/blob/nullable-projection/src/IntegrationsTests.EFCore/ProjectionConversionTests.cs

**Reply by lbargaoanu** [reply]
Yes, because custom conversion operators are for `Map`, not for `ProjectTo`.

**Reply by rmja** [reply]
That is most likely correct, I do not have a deep understanding of AutoMapper, but my point is: Would it somehow be possible for `ProjectTo` to do "some kind of implicit conversion" when implicit operators exist for the type?

**Reply by lbargaoanu** [reply]
No, because that would be client evaluation, that's what `Map` does.

**Reply by rmja** [reply]
So what you are saying is that I am doing client evaluation when I do `ProjectTo()` with:
```
config.CreateMap<SourceValue, DestinationValue>().ConvertUsing(src => src);
config.CreateMap<SourceValue?, DestinationValue?>().ConvertUsing(src => src);
```

?

**Reply by lbargaoanu** [reply]
The custom operator is client evaluation. That other thing is some unnatural act :) Check [the execution plan](https://docs.automapper.org/en/latest/Understanding-your-mapping.html).
